### PR TITLE
Add support for ECDSA P256 using mbed TLS

### PR DESCRIPTION
**Pull request by d3zd3z**
Although tinycrypt is smaller and faster than mbed TLS at ECDSA, when some other functionality needs mbed TLS, bringing in tinycrypt as well would only serve to make the code larger. This patchset adds support for mbed TLS with ecdsa.

### CODE REVIEW

**Comment by d3zd3z** [comment]
The abstraction ends up rather messy from all of this, and I'll probably end up redoing how it is done.

**Comment by d3zd3z** [comment]
More in-depth discussion in #922.

**Comment by d3zd3z** [comment]
I think this is a reasonable compromise solution. I do have a single define that affects the calling code as to whether it wants the key decoded or not. Before, I was decoding the key and then possibly not using the decoded key, which is less efficient.

File: boot/bootutil/src/image_ec256.c [hunk 1]
```diff
@@ -30,10 +30,11 @@
 
 #ifdef MCUBOOT_SIGN_EC256
 /*TODO: remove this after cypress port mbedtls to abstract crypto api */
-#ifdef MCUBOOT_USE_CC310
+#if defined(MCUBOOT_USE_CC310) || defined(MCUBOOT_USE_MBED_TLS)
 #define NUM_ECC_BYTES (256 / 8)
 #endif
-#if defined (MCUBOOT_USE_TINYCRYPT) || defined (MCUBOOT_USE_CC310)
+#if defined(MCUBOOT_USE_TINYCRYPT) || defined(MCUBOOT_USE_CC310) || \
+	defined(MCUBOOT_USE_MBED_TLS)
```
Comment (utzig) [inline boot/bootutil/src/image_ec256.c:37]: tabs -> spaces

File: boot/bootutil/include/bootutil/crypto/ecdsa_p256.h [hunk 2]
```diff
@@ -69,13 +86,77 @@ static inline void bootutil_ecdsa_p256_drop(bootutil_ecdsa_p256_context *ctx)
...
+
+static inline void bootutil_ecdsa_p256_drop(bootutil_ecdsa_p256_context *ctx)
+{
+    mbedtls_ecdsa_free(ctx);
+}
+
+static inline int bootutil_ecdsa_p256_verify(bootutil_ecdsa_p256_context *ctx,
+					     uint8_t *pk, size_t pk_len,
+					     uint8_t *hash,
+					     uint8_t *sig, size_t sig_len)
```
Comment (utzig) [inline boot/bootutil/include/bootutil/crypto/ecdsa_p256.h:128]: tabs -> spaces

File: boot/bootutil/include/bootutil/crypto/ecdsa_p256.h [hunk 3]
```diff
@@ -69,13 +86,77 @@ static inline void bootutil_ecdsa_p256_drop(bootutil_ecdsa_p256_context *ctx)
     (void)ctx;
 }
 
-static inline int bootutil_ecdsa_p256_verify(bootutil_ecdsa_p256_context *ctx, uint8_t *pk, uint8_t *hash, uint8_t *sig)
+static inline int bootutil_ecdsa_p256_verify(bootutil_ecdsa_p256_context *ctx,
+					     uint8_t *pk, size_t pk_len,
+					     uint8_t *hash,
+					     uint8_t *sig, size_t sig_len)
```
Comment (utzig) [inline boot/bootutil/include/bootutil/crypto/ecdsa_p256.h:92]: tabs -> spaces

**Review by utzig — CHANGES_REQUESTED** [review]
Implementation looks fine, would be nice to also have #884 for a complete Mbed TLS EC based implementation.

File: boot/zephyr/include/config-ecdsa.h [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+#define MBEDTLS_PK_PARSE_C
+#define MBEDTLS_X509_CRT_PARSE_C
+
+/* Save RAM by adjusting to our exact needs */
+#define MBEDTLS_ECP_MAX_BITS             256
+
+#if (CONFIG_BOOT_SIGNATURE_TYPE_RSA_LEN == 3072)
+#define MBEDTLS_MPI_MAX_SIZE              384
+#else
+#define MBEDTLS_MPI_MAX_SIZE              256
```
[the remark this answers is not in the document]
Comment (d3zd3z) [reply boot/zephyr/include/config-ecdsa.h:81]: Actually, these values don't really make any sense for ECDSA. 32 is fine here, and will save RAM.

**Comment by d3zd3z** [comment]
@utzig: I've fixed the tabs, and replaced the MPI size with just 32, which is the value needed for P256.

**Review by utzig — APPROVED** [review]
LGTM

**Review by nvlsianpu — APPROVED** [review]

**Comment by nvlsianpu** [comment]
Do  anyone know why `Travis CI - Branch` workflow was triggered?

**Comment by utzig** [comment]
> Do anyone know why `Travis CI - Branch` workflow was triggered?

A branch was created on this fork and subsequently deleted.

**Comment by d3zd3z** [comment]
> > Do anyone know why `Travis CI - Branch` workflow was triggered?
> 
> A branch was created on this fork and subsequently deleted.

Yeah, my bad. I accidentally pushed to the main project instead of my own repo. I deleted it shortly thereafter, but it was there long enough to anger Travis.

**Pull request merged**